Add GetNodeVersionAsync() to get full node UA string for settings

prayank23 opened the pull request:
Fixes https://github.com/zkSNACKs/WalletWasabi/issues/6785


<details>
  <summary>With full node</summary>
  
![image](https://user-images.githubusercontent.com/13405205/147377966-2b76ff77-ce95-49dd-bbfb-3680e9e9f47c.png)

</details>

<details>
  <summary>Without full node</summary>
  
![image](https://user-images.githubusercontent.com/13405205/147380456-6b965c4a-7649-4110-b814-de901f00abec.png)

</details>

prayank23 commented:
One test is failing since I added new method to get version. I am not sure how to fix this or if its required to test this: `WalletWasabi.Tests/UnitTests/BitcoinCore/NodeBuildingTests.cs#L78 `

lontivero commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -115,5 +117,22 @@ protected override void EditConfigOnSave(Config config)
+			var version = Constants.BitcoinCoreVersion;
+			try
+			{
+				var node = Node.Connect(Services.Config.Network, Services.Config.GetBitcoinP2pEndPoint().ToString(defaultPort: -1));
```
Prefer non-blocking calls. In this case `Node.ConnectAsync`. Also remember `Node` is disposable and we dispose them by using `using var node = XXXX`

lontivero replied:
the `ToString(defaultPortt: -1)` is unnecessary. Remove it, please.

prayank23 replied:
Done in https://github.com/zkSNACKs/WalletWasabi/pull/6942/commits/af03334456ed50f28b622d47f0d15d0e8c3a428f

lontivero reviewed (changes requested):
Looks good. Fix the tests please.

kiminuo commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -54,7 +56,7 @@ public BitcoinTabSettingsViewModel()
 		}
 
-		public Version BitcoinCoreVersion => Constants.BitcoinCoreVersion;
+		public string BitcoinCoreVersion => GetNodeVersion();
```
This property, supposedly, calls `GetNodeVersion` every time the settings dialog is opened. This in turn allows us to have always a correct bitcoind version (i.e. one can terminate `bitcoind` and run a newer version and we would report a correct new value).

I don't think it can noticably slow down displaying of the settings page. If it somehow can, we should probably resort to caching.

prayank23 replied:
> I don't think it can noticably slow down displaying of the settings page. If it somehow can, we should probably resort to caching.

I don't think its required.

kiminuo commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -115,5 +117,22 @@ protected override void EditConfigOnSave(Config config)
 			}
 		}
+
+		public static string GetNodeVersion()
```
I would rather return `null` when we can't get a version string (if bitcoind fails to give us result for whatever reason) and that `null` should turn into some user-friendly message for the UI user.

kiminuo commented on WalletWasabi/Helpers/Constants.cs:
```diff
@@ -55,7 +55,7 @@ public static class Constants
 		public static readonly Version ClientVersion = new(1, 1, 13, 0);
 		public static readonly Version HwiVersion = new("2.0.2");
-		public static readonly Version BitcoinCoreVersion = new("0.21.1");
+		public static readonly string BitcoinCoreVersion = new("");
```
This feels weird to me.

prayank23 replied:
Reverted in https://github.com/zkSNACKs/WalletWasabi/pull/6942/commits/af03334456ed50f28b622d47f0d15d0e8c3a428f

yahiheb commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -115,5 +117,22 @@ protected override void EditConfigOnSave(Config config)
 			}
 		}
+
+		public static string GetNodeVersion()
+		{
+			var version = Constants.BitcoinCoreVersion;
```
```suggestion
		private string GetNodeVersion()
		{
			string version = new("");
```

yahiheb reviewed (changes requested):
@prayank23 I suppose what you want to achieve is to display the correct local full node version that Wasabi is using in the Settings.

To do that without changing the supported compatible full node version in the About dialog I propose the following:

- Revert changes in: `AboutAdvancedInfoViewModel.cs` & `WalletWasabi/Helpers/Constants.cs`
- Apply suggestion in: `BitcoinTabSettingsViewModel.cs`

This will make the test pass again.

prayank23 commented:
Thanks @yahiheb @lontivero and @kiminuo for reviewing pull request

Addressed most of the comments above from review:

- Revert constant used for Bitcoin Full node
- Revert to constant in advanced view
- Change binding variable and add `^` according to https://docs.avaloniaui.net/docs/data-binding/binding-to-tasks-and-observables#example-2-binding-to-a-task
- Make method async

yahiheb commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -54,7 +57,7 @@ public BitcoinTabSettingsViewModel()
 		}
 
-		public Version BitcoinCoreVersion => Constants.BitcoinCoreVersion;
+		public Task<string> BitcoinKnotsVersion => GetNodeVersionAsync();
```
How about renaming the variable to:

```suggestion
		public Task<string> BitcoinFullNodeVersion => GetNodeVersionAsync();
```

prayank23 replied:
This variable name won't be visible to users who do not read code. It was Core earlier even though we support Knots. Core is still used at lot of places so its okay to use Knots here.

yahiheb replied:
This is not about the built-in node binary that we ship, if that was the case then there is no reason for this PR in the first place and hard coding the version is enough.
This represents the full node version that the user is using, it might be Core or Knots or anything else.

> It was Core earlier even though we support Knots. Core is still used at lot of places so its okay to use Knots here.

If naming the variable with Core is wrong then it should be fixed, not using Core with some variables and Knots with other ones. Consistency.

prayank23 replied:
There is no consistency in the name of variables used for full node.

yahiheb replied:
No need to add more inconsistency then.

prayank23 replied:
I prefer Bitcoin Knots in this variable name. You are free to NACK the PR or open a separate PR.

prayank23 replied:
Done

yahiheb commented on WalletWasabi.Fluent/Views/Settings/BitcoinTabSettingsView.axaml:
```diff
@@ -27,8 +27,8 @@
-      <TextBlock VerticalAlignment="Center" Text= "Local Bitcoin Core/Knots version" />
-      <Label Content="{Binding BitcoinCoreVersion}" />
+      <TextBlock VerticalAlignment="Center" Text= "Local Bitcoin Full Node " />
+      <Label Content="{Binding BitcoinKnotsVersion^}" />
```
Same:

```suggestion
      <Label Content="{Binding BitcoinFullNodeVersion^}" />
```

prayank23 replied:
Done

molnard commented:
Concept ACK

adamPetho commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -115,5 +118,22 @@ protected override void EditConfigOnSave(Config config)
+				using var node = await Node.ConnectAsync(Services.Config.Network, Services.Config.GetBitcoinP2pEndPoint());
+				node.VersionHandshake();
+				version = node.PeerVersion.UserAgent;
+				Logger.LogDebug($"Full Node Version: {version} ");
```
IMO this line of logging is unnecessary.

prayank23 replied:
Removed

adamPetho commented on WalletWasabi.Fluent/ViewModels/Settings/BitcoinTabSettingsViewModel.cs:
```diff
@@ -115,5 +118,22 @@ protected override void EditConfigOnSave(Config config)
+			}
+			catch (Exception ex)
+			{
+				Logger.LogDebug(ex);
```
```suggestion
				Logger.LogError(ex);
```
IMO we want to see in the logs, when we couldn't even get our Full Node's version.

prayank23 replied:
Done

adamPetho reviewed (commented):
Fix the merge conflict please, also nitpicks below.

adamPetho commented:
Also, without a Full Node, the settings is just `Local Bitcoin Full Node` then nothing, only emptiness.
How about we don't show that line if we don't have a full node?

I think it could look something like this [NOT TESTED]:

```
public async Task<bool> TryGetNodeVersionAsync([NotNullWhenTrue] out string version)
{
	try
	{
		using var node = await Node.ConnectAsync(Services.Config.Network, Services.Config.GetBitcoinP2pEndPoint());
		node.VersionHandshake();
		version = node.PeerVersion.UserAgent;
                return true;
	}
	catch (Exception ex)
	{
		Logger.LogDebug(ex);
	}
	return false;
}
```

Then in the ViewModel we could have
```
[AutoNotify] private string? _fullNodeVersion;
[AutoNotify] private bool _showFullNodeVersion;
```

in the `ctor` we could call the `TryGetNodeVersionAsync` and fill up the 2 new prop accordingly

Then the `axaml` file could look like this:

```
<TextBlock VerticalAlignment="Center" Text= "Bitcoin Node Version " IsVisible="{Binding ShowFullNodeVersion}" />
      <Label Content="{Binding FullNodeVersion^}" />
```

prayank23 commented:
> Also, without a Full Node, the settings is just Local Bitcoin Full Node then nothing, only emptiness.
How about we don't show that line if we don't have a full node?

@adamPetho 

<details>
  <summary>PR Branch (when no local full node is running)</summary>

![image](https://user-images.githubusercontent.com/13405205/152421831-23c1fcce-3be1-4876-aff2-fbb6fffe1589.png)

![image](https://user-images.githubusercontent.com/13405205/152421873-42a4dfa6-1514-454b-a298-0db190a8b035.png)
  
</details>


<details>
  <summary>Your suggestion (when no local full node is running)</summary>

![image](https://user-images.githubusercontent.com/13405205/152422087-7fbe7db4-8860-49fb-93b2-933dd91f5991.png)

![image](https://user-images.githubusercontent.com/13405205/152422131-af9fa282-a23c-49f1-bc86-2b68a7a47586.png)

  
</details>

prayank23 commented:
@adamPetho this partially fixes the issue that you mentioned and will not print label if version string is empty:

```axml
    <DockPanel>
      <TextBlock VerticalAlignment="Center" Text= "Local Bitcoin Full Node " IsVisible="{Binding BitcoinNodeVersion^, Converter={x:Static StringConverters.IsNotNullOrEmpty}}" />
      <Label Content="{Binding BitcoinNodeVersion^}" />
    </DockPanel>
```

adamPetho commented:
@prayank23  I wonder if we even need the `BitcoinNodeVersion^` binding or not if we use that `StringConverter`
